exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -317,3 +324,259 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+		return r.reconcileFallbackMachines(ctx, mp)
+	} else if noKind != noSelector {
+		log.Info("Only one of infraMachineKind or infraMachineSelector found, this is unexpected")
+		return errors.Errorf("only one of infraMachineKind or infraMachineSelector found, both must be present")
+	}
+
+	log.Info("Reconciling MachinePool Machines", "infrastructureMachineKind", infraMachineKind, "infrastructureMachineSelector", infraMachineSelector)
+	var infraMachineList unstructured.UnstructuredList
+
+	infraMachineList.SetAPIVersion(infraMachinePool.GetAPIVersion())
+	infraMachineList.SetKind(infraMachineKind)
+
+	if err := r.Client.List(ctx, &infraMachineList, client.InNamespace(mp.Namespace), client.MatchingLabels(infraMachineSelector.MatchLabels)); err != nil {
+		return errors.Wrapf(err, "failed to list infra machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	if err := r.deleteDanglingMachines(ctx, mp); err != nil {
+		return errors.Wrapf(err, "failed to clean up orphan machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	if err := r.createMachinesIfNotExists(ctx, mp, infraMachineList.Items); err != nil {
+		return errors.Wrapf(err, "failed to create machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	return nil
+}
+
+// deleteDanglingMachines deletes all MachinePool Machines with an infraRef pointing to an infraMachine that no longer exists.
+// This allows us to clean up Machines when the MachinePool is scaled down.
+func (r *MachinePoolReconciler) deleteDanglingMachines(ctx context.Context, mp *expv1.MachinePool) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Deleting orphaned machines", "machinePool", mp.Name, "namespace", mp.Namespace)
+	machineList := &clusterv1.MachineList{}
+	labels := map[string]string{
+		expv1.MachinePoolNameLabel:      mp.Name,
+		clusterv1.ClusterNameLabel:      mp.Spec.ClusterName,
+		clusterv1.MachinePoolOwnedLabel: "true",
+	}
+
+	if err := r.Client.List(ctx, machineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
+		return err
+	}
+
+	for i := range machineList.Items {
+		machine := &machineList.Items[i]
+		infraRef := machine.Spec.InfrastructureRef
+
+		// Check this here since external.Get() will complain if the infraRef is empty, i.e. no object kind or apiVersion.
+		if (infraRef == corev1.ObjectReference{}) {
+			log.V(2).Info("Machine has an empty infraRef, this is likely a fallback machine, will delete anyway", "machine", machine.Name, "namespace", machine.Namespace)
+
+			return r.Client.Delete(ctx, machine)
+		}
+
+		_, err := external.Get(ctx, r.Client, &infraRef, mp.Namespace)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Error(err, "unexpected error while trying to get infra machine, will delete anyway", "machine", machine.Name, "namespace", machine.Namespace)
+			}
+
+			log.V(2).Info("Deleting orphaned machine", "machine", machine.Name, "namespace", machine.Namespace)
```

Comment:
I wonder if this should be a higher log level (v1), or do we expect it to be noisy?

Reply:
This would run whenever a MachinePool is scaled down, do you think that would be too noisy? If not, we can make it V(1).